fpdf/text_region.py

```diff
@@ -106,6 +136,30 @@ def write(self, text: str, link=None):
                 frag.link = link
         self._text_fragments.extend(fragments)
 
+    def generate_bullet_frag_and_tl(
+        self, bullet_string: str, rel_x_displacement: float
+    ):
+        if not bullet_string:
+            return None
+        if not self.pdf.font_family:
+            raise FPDFException("No font set, you need to call set_font() beforehand")
+        (bullet_fragment,) = self.pdf._preload_font_styles(
```

Comment:
Tested with `<ul>` bullet strings '😁' and 'aaaaa aaaaa😁' (by editing the variable during debug) using DejaVuSans as the font. Found no issues - this seems to be working as expected.

Given that there is already `test_html_ul_type` with a test for `<ul type="■">` using DejaVuSans, I take it that there is no need for introducing another test, is there?

Reply:
DejaVuSans actually includes glyphs for all those characters, so that's not really a surprise...

You'll have to be a bit more creative with your testing and find a more obscure combination! 😜

Reply:
I initially thought that you wanted me to make sure that the cases that worked previously work fine after my changes.

I added `self.pdf.normalize_text(bullet_string)` to handle these issues.

Reply:
> I initially thought that you wanted me to make sure that the cases that worked previously work fine after my changes.

That is the fundamental requirement for any change. But if you add features, then you also add new use cases, all of which should get tested as thoroughly as reasonably possible.
 
> I added `self.pdf.normalize_text(bullet_string)` to handle these issues.

And you assumed that it would only return a single-item tuple in all your use cases, without testing this assumption carefully enough.

> Also, not sure how there can be such a case where there are multiple fragments generated by `self.pdf._preload_font_styles`, as the only relevant custom bullet option seems to be the `type` attribute of `<ul>`, but that one is (at least currently) limited to just one character.

As I just wrote with another point nearby: Parsing HTML is only one of many different use cases for bulleted/numbered paragraphs. Of course that one should also work, but ignoring all the others is a receipe for receiving heaps of bug reports later.